src/lib/build.js

```diff
@@ -55,10 +66,52 @@ const runBuild = async (options) => {
       scheme: apiUrl.protocol.slice(0, -1),
       host: apiUrl.host,
     }
-    options = { ...options, testOpts }
+    buildOptions = { ...buildOptions, testOpts }
+  }
+
+  /** @type {Partial<import('../../utils/types').ServerSettings>} */
+  let settings = {}
+  try {
+    settings = await detectServerSettings(devConfig, commandOptions, site.root)
+
+    const defaultConfig = { build: {} }
+
+    if (settings.buildCommand) {
+      buildOptions.cachedConfig.config.build.command = settings.buildCommand
+      defaultConfig.build.command = settings.buildCommand
+    }
+    if (settings.dist) {
+      buildOptions.cachedConfig.config.build.publish = settings.buildCommand
+      defaultConfig.build.publish = settings.dist
+    }
+
+    buildOptions.defaultConfig = defaultConfig
+
+    // If there are plugins that we should be running for this site, add them
+    // to the config as if they were declared in netlify.toml. We must check
+    // whether the plugin has already been added by another source (like the
+    // TOML file or the UI), as we don't want to run the same plugin twice.
+    if (settings.plugins) {
+      const { plugins: existingPlugins = [] } = cachedConfig.config
+      const existingPluginNames = new Set(existingPlugins.map((plugin) => plugin.package))
+      const newPlugins = settings.plugins
+        .map((pluginName) => {
+          if (existingPluginNames.has(pluginName)) {
+            return
+          }
+
+          return { package: pluginName, origin: 'config', inputs: {} }
+        })
+        .filter(Boolean)
+
+      buildOptions.cachedConfig.config.plugins = [...newPlugins, ...cachedConfig.config.plugins]
+    }
+  } catch (error_) {
+    log(NETLIFYDEVERR, error_.message)
+    exit(1)
```

Comment:
I believe the `error` util should handle this?

```suggestion
    error(error_)
```